### PR TITLE
Allow components as userhooks

### PR DESCRIPTION
**Pull request by stdweird**
Some configuration modules are providing the support for operations that must be done during the installation (AII), e.g. `ncm-opennebula` or `ncm-freeipa`.

Depends on #202

### CODE REVIEW

**Comment by stdweird** [comment]
We just discovered a possible  naming issue. the components have a `Configure` method, the hooks possibly have a `configure` method. (to make this more confusing, the plugins also have a `Configure` method, so we cannot use components as plugins, only as hooks).

question is: should we rename the hook methods for Components to e.g. `aii_install` (and also `aii_configure`)? if we do, there's not naming confict/ugliness, and we could use components as plugins too

**Comment by jouvin** [comment]
Agreement at the workshop with the proposal of an entry point dedicated to the use of the component in the AII context and this entry point being the only one used in the NCM context.

File: aii-pxelinux/src/main/pan/quattor/aii/pxelinux/config.pan [hunk 1]
```diff
@@ -56,26 +59,20 @@ variable AII_NBP_LABEL ?= {
...
-            label = 'Scientific Linux CERN ';
-        } else if ( toks[1] == 'rhel' ) {
-            label = 'Red Hat Entreprise Linux ';
-        } else {
-            label = undef;
-        };
-        if ( is_defined(label) ) {
-            label = label + toks[2] + ' ('+ toks[3] + ')';
-        };
+        label = format("%s %s (%s)", labelmap[toks[1]], toks[2], toks[3]);
```
Comment (piojo) [inline aii-pxelinux/src/main/pan/quattor/aii/pxelinux/config.pan]: Thanks! :+1:

File: aii-pxelinux/src/main/pan/quattor/aii/pxelinux/schema.pan [hunk 2]
```diff
@@ -2,9 +2,14 @@
...
 # ${build-info}
-unique template quattor/aii/pxelinux/schema;
 
-# PXE configuration.
+declaration template quattor/aii/pxelinux/schema;
+
+include 'pan/types';
+
+@documentation{
+    PXE configuration
```
Comment (piojo) [inline aii-pxelinux/src/main/pan/quattor/aii/pxelinux/schema.pan]: Give me some more details, please :smile:

File: aii-pxelinux/src/main/perl/pxelinux.pm [hunk 3]
```diff
@@ -45,20 +45,20 @@ use constant KS => "/system/aii/osinstall/ks";
 # Lowest supported version is EL 5.0
 use constant ANACONDA_VERSION_EL_5_0 => version->new("11.1");
 use constant ANACONDA_VERSION_EL_6_0 => version->new("13.21");
-use constant ANACONDA_VERSION_EL_7_0 => version->new("19.31"); 
+use constant ANACONDA_VERSION_EL_7_0 => version->new("19.31");
 use constant ANACONDA_VERSION_LOWEST => ANACONDA_VERSION_EL_5_0;
```
Comment (piojo) [inline aii-pxelinux/src/main/perl/pxelinux.pm]: ???

What a cumbersome way to say

```perl
if (!exists($tree->{bonding})) {
    $this_app->verbose("Bonding config generation not defined...");
} elsif ($bonding_disabled) {
    $this_app->verbose("Bonding configuration explicitly disabled");
}
```

File: aii-pxelinux/src/main/perl/pxelinux.pm [hunk 4]
```diff
@@ -115,27 +115,27 @@ sub link_filepath
 }
```
Comment (piojo) [inline aii-pxelinux/src/main/perl/pxelinux.pm]: Why cannot you handle all the `$bonding_disabled` cases together?

File: aii-core/src/main/perl/Shellfe.pm [hunk 5]
```diff
@@ -358,54 +358,56 @@ sub run_plugin
...
+
     # This is here because CacheManager and Fetch objects may have
     # problems when they get out of scope.
     my %rm = $st->{configuration}->getElement ($path)->getHash;
-    my $modulename = (keys (%rm))[0];
-    if ($modulename !~ m/^[a-zA-Z_]\w+$/) {
-        $self->error ("Invalid Perl identifier specified as a plug-in. ",
-                      "Skipping.");
+    my $modulename = (sort keys (%rm))[0];
+    if ($modulename !~ m/^[a-zA-Z_]\w+(::[a-zA-Z_]\w+)*$/) {
```
Comment (piojo) [inline aii-core/src/main/perl/Shellfe.pm]: Agreed in the workshop that this would be acceptable

**Comment by stdweird** [comment]
added `aii_` prefix for methods in `NCM::Component` namespace

**Comment by jrha** [comment]
LGTM AFAICT

**Comment by jrha** [comment]
@stdweird I'm not sure @piojo's comments apply any more, but could you check quickly?

**Comment by stdweird** [comment]
@jrha some of @piojo remarks were about #202 . i rebased this PR, so they are not relevant anymore

**Comment by jrha** [comment]
Ok thanks.

**Pull request merged**